libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance.hs

```diff
@@ -1,10 +1,455 @@
+
+  toSpecRep (ValidityInterval lo hi) = (,) <$> toSpecRep lo <*> toSpecRep hi
+
+instance SpecCastable (Hash a b) where
+  type SpecRep (Hash a b) = Agda.Hash
+
+  toSpecRep = pure . byteStringToInteger . hashToBytes
+
+instance SpecCastable (KeyHash r c) where
+  type SpecRep (KeyHash r c) = Agda.Hash
+
+  toSpecRep (KeyHash h) = toSpecRep h
+
+instance
+  Crypto c =>
+  SpecCastable (VKey k c)
+  where
+  type SpecRep (VKey k c) = Integer
+
+  toSpecRep (VKey x) = pure . byteStringToInteger $ rawSerialiseVerKeyDSIGN x
+
+instance DSIGNAlgorithm v => SpecCastable (SignedDSIGN v a) where
+  type SpecRep (SignedDSIGN v a) = Integer
+
+  toSpecRep (SignedDSIGN x) =
+    pure . byteStringToInteger $ rawSerialiseSigDSIGN x
+
+instance (Crypto c, Typeable k) => SpecCastable (WitVKey k c) where
+  type SpecRep (WitVKey k c) = (Integer, Integer)
+
+  toSpecRep (WitVKey vk sk) =
+    (,)
+      <$> toSpecRep vk
+      <*> toSpecRep sk
+
+instance Era era => SpecCastable (TxDats era) where
+  type SpecRep (TxDats era) = [(Agda.DataHash, Agda.Datum)]
+
+  toSpecRep (TxDats x) = toSpecRep x
+
+instance (SpecCastable k, SpecCastable v) => SpecCastable (Map k v) where
+  type SpecRep (Map k v) = [(SpecRep k, SpecRep v)]
+
+  toSpecRep m = traverse act $ Map.toList m
+    where
+      act (k, v) = do
+        k' <- toSpecRep k
+        v' <- toSpecRep v
+        pure (k', v')
+
+instance SpecCastable Tag where
+  type SpecRep Tag = Agda.Tag
+
+  toSpecRep Spend = pure Agda.Spend
+  toSpecRep Mint = pure Agda.Mint
+  toSpecRep Cert = pure Agda.Cert
+  toSpecRep Rewrd = pure Agda.Rewrd
+
+instance SpecCastable Word64 where
+  type SpecRep Word64 = Integer
+
+  toSpecRep = pure . toInteger
+
+instance SpecCastable RdmrPtr where
+  type SpecRep RdmrPtr = Agda.RdmrPtr
+
+  toSpecRep (RdmrPtr t x) = toSpecRep (t, x)
+
+instance (SpecCastable a, SpecCastable b) => SpecCastable (a, b) where
+  type SpecRep (a, b) = (SpecRep a, SpecRep b)
+
+  toSpecRep (x, y) = (,) <$> toSpecRep x <*> toSpecRep y
+
+instance SpecCastable (Data era) where
+  type SpecRep (Data era) = ()
+
+  toSpecRep _ = pure ()
+
+instance Era era => SpecCastable (Redeemers era) where
+  type SpecRep (Redeemers era) = [(Agda.RdmrPtr, (Agda.Redeemer, Agda.ExUnits))]
+
+  toSpecRep (Redeemers x) = toSpecRep x
+
+instance
+  ( Era era
+  , Script era ~ AlonzoScript era
+  ) =>
+  SpecCastable (AlonzoTxWits era)
+  where
+  type SpecRep (AlonzoTxWits era) = Agda.TxWitnesses
+
+  toSpecRep x =
+    Agda.MkTxWitnesses
+      <$> toSpecRep (txwitsVKey x)
+      <*> pure []
+      <*> toSpecRep (txdats x)
+      <*> toSpecRep (txrdmrs x)
+
+instance SpecCastable a => SpecCastable (StrictMaybe a) where
+  type SpecRep (StrictMaybe a) = Maybe (SpecRep a)
+
+  toSpecRep = traverse toSpecRep . strictMaybeToMaybe
+
+instance SpecCastable (AlonzoTxAuxData era) where
+  type SpecRep (AlonzoTxAuxData era) = Agda.AuxiliaryData
+
+  toSpecRep _ = pure ()
```

Comment:
I think it's safer to succeed here and then if we can't manage to make the tests pass, then we can fail in these places. Because if we fail here, we discard tests, while if we succeed with the translation, we get a test failure.

Reply:
Sure. If you think it makes sense to postpone making this decision, I have nothing against it